File: backend-api/app/schemas/instrument.py
"""
仪器 Pydantic 模型

定义仪器相关的请求和响应模型，包括数据验证规则。
"""
from pydantic import BaseModel, Field, field_validator
from typing import Optional, Dict, Any
from datetime import datetime
from enum import Enum

from app.schemas.response import PaginationInfo


class InstrumentStatus(str, Enum):
    """仪器状态枚举"""
    IN_USE = "IN_USE"
    STANDBY = "STANDBY"
    MAINTENANCE = "MAINTENANCE"
    CALIBRATING = "CALIBRATING"
    PENDING_DISPOSAL = "PENDING_DISPOSAL"
    DISPOSED = "DISPOSED"


class InstrumentBase(BaseModel):
    """仪器基础模型"""
    name: str = Field(..., min_length=1, max_length=200, description="仪器名称")
    model: Optional[str] = Field(None, max_length=100, description="型号")
    manufacturer: Optional[str] = Field(None, max_length=200, description="制造商")
    serial_number: Optional[str] = Field(None, max_length=100, description="序列号")
    purchase_date: Optional[datetime] = Field(None, description="购置日期")
    purchase_price: Optional[float] = Field(None, ge=0, description="购置价格")
    technical_params: Optional[Dict[str, Any]] = Field(None, description="技术参数")
    current_location: Optional[str] = Field(None, max_length=200, description="当前位置")
    current_department: Optional[str] = Field(None, max_length=200, description="当前部门")
    current_responsible: Optional[str] = Field(None, max_length=100, description="当前负责人")
    usage_years: Optional[int] = Field(None, ge=0, description="使用年限")
    warranty_expiry: Optional[datetime] = Field(None, description="保修到期日期")
    description: Optional[str] = Field(None, description="描述")
    remarks: Optional[str] = Field(None, description="备注")

    @field_validator('name', 'model', 'manufacturer', 'serial_number', mode='before')
    @classmethod
    def strip_strings(cls, v):
        """清洗字符串，去除首尾空格"""
        if isinstance(v, str):
            return v.strip()
        return v

    @field_validator('current_location', 'current_department', 'current_responsible', 
                     'description', 'remarks', mode='before')
    @classmethod
    def strip_optional_strings(cls, v):
        """清洗可选字符串字段"""
        if isinstance(v, str):
            stripped = v.strip()
            return stripped if stripped else None
        return v


class InstrumentCreate(InstrumentBase):
    """创建仪器请求模型"""
    code: str = Field(..., min_length=1, max_length=100, description="仪器编码")
    status: InstrumentStatus = Field(default=InstrumentStatus.IN_USE, description="状态")
    
    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "code": "INS-2024-001",
                    "name": "高效液相色谱仪",
                    "model": "LC-2030C",
                    "manufacturer": "岛津",
                    "serial_number": "C12345678",
                    "purchase_date": "2024-01-15T00:00:00",
                    "purchase_price": 350000.0,
                    "technical_params": {
                        "measurementRange": "190-800nm",
                        "precision": "±0.5%",
                        "resolution": "0.1nm"
                    },
                    "status": "IN_USE",
                    "current_location": "检测室A",
                    "current_department": "理化检测部",
                    "current_responsible": "张三",
                    "description": "用于水质检测",
                    "remarks": "新购设备"
                }
            ]
        }
    }


class InstrumentUpdate(BaseModel):
    """更新仪器请求模型（所有字段可选）"""
    name: Optional[str] = Field(None, min_length=1, max_length=200, description="仪器名称")
    model: Optional[str] = Field(None, max_length=100, description="型号")
    manufacturer: Optional[str] = Field(None, max_length=200, description="制造商")
    serial_number: Optional[str] = Field(None, max_length=100, description="序列号")
    purchase_date: Optional[datetime] = Field(None, description="购置日期")
    purchase_price: Optional[float] = Field(None, ge=0, description="购置价格")
    technical_params: Optional[Dict[str, Any]] = Field(None, description="技术参数")
    status: Optional[InstrumentStatus] = Field(None, description="状态")
    current_location: Optional[str] = Field(None, max_length=200, description="当前位置")
    current_department: Optional[str] = Field(None, max_length=200, description="当前部门")
    current_responsible: Optional[str] = Field(None, max_length=100, description="当前负责人")
    usage_years: Optional[int] = Field(None, ge=0, description="使用年限")
    warranty_expiry: Optional[datetime] = Field(None, description="保修到期日期")
    description: Optional[str] = Field(None, description="描述")
    remarks: Optional[str] = Field(None, description="备注")

    @field_validator('name', 'model', 'manufacturer', 'serial_number', mode='before')
    @classmethod
    def strip_strings(cls, v):
        """清洗字符串，去除首尾空格"""
        if isinstance(v, str):
            return v.strip()
        return v

    @field_validator('current_location', 'current_department', 'current_responsible', 
                     'description', 'remarks', mode='before')
    @classmethod
    def strip_optional_strings(cls, v):
        """清洗可选字符串字段"""
        if isinstance(v, str):
            stripped = v.strip()
            return stripped if stripped else None
        return v
    
    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "current_location": "检测室B",
                    "current_department": "微生物检测部",
                    "status": "MAINTENANCE",
                    "remarks": "设备维护中"
                }
            ]
        }
    }


class InstrumentResponse(InstrumentBase):
    """仪器响应模型"""
    id: str = Field(..., description="仪器 ID")
    code: str = Field(..., description="仪器编码")
    status: InstrumentStatus = Field(..., description="状态")
    version: int = Field(..., description="版本号")
    created_by: str = Field(..., description="创建人")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")
    
    model_config = {
        "from_attributes": True,
        "json_schema_extra": {
            "examples": [
                {
                    "id": "550e8400-e29b-41d4-a716-446655440000",
                    "code": "INS-2024-001",
                    "name": "高效液相色谱仪",
                    "model": "LC-2030C",
                    "manufacturer": "岛津",
                    "serial_number": "C12345678",
                    "purchase_date": "2024-01-15T00:00:00",
                    "purchase_price": 350000.0,
                    "technical_params": {
                        "measurementRange": "190-800nm",
                        "precision": "±0.5%",
                        "resolution": "0.1nm"
                    },
                    "status": "IN_USE",
                    "current_location": "检测室A",
                    "current_department": "理化检测部",
                    "current_responsible": "张三",
                    "usage_years": 0,
                    "warranty_expiry": "2027-01-15T00:00:00",
                    "description": "用于水质检测",
                    "remarks": "新购设备",
                    "version": 1,
                    "created_by": "550e8400-e29b-41d4-a716-446655440001",
                    "created_at": "2024-01-15T09:00:00",
                    "updated_at": "2024-01-15T09:00:00"
                }
            ]
        }
    }


class InstrumentListResponse(BaseModel):
    """仪器列表响应模型"""
    items: list[InstrumentResponse] = Field(..., description="仪器列表")
    pagination: PaginationInfo = Field(..., description="分页信息")


class InstrumentStatusUpdate(BaseModel):
    """仪器状态更新请求模型"""
    status: InstrumentStatus = Field(..., description="新状态")
    
    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "status": "MAINTENANCE"
                }
            ]
        }
    }

File: backend-api-deprecated/src/types/instrument.ts
/**
 * 仪器管理相关类型定义
 */

// ============================================
// 枚举类型 (与 Prisma Schema 保持一致)
// ============================================

/**
 * 仪器状态枚举
 */
export enum InstrumentStatus {
  IN_USE = 'IN_USE',                    // 在用
  STANDBY = 'STANDBY',                  // 备用
  MAINTENANCE = 'MAINTENANCE',          // 维修中
  CALIBRATING = 'CALIBRATING',          // 校准中
  PENDING_DISPOSAL = 'PENDING_DISPOSAL', // 待报废
  DISPOSED = 'DISPOSED'                 // 已报废
}

/**
 * 流转状态枚举
 */
export enum InstrumentTransferStatus {
  PENDING = 'PENDING',      // 待确认
  CONFIRMED = 'CONFIRMED',  // 已确认
  REJECTED = 'REJECTED',    // 已拒绝
  COMPLETED = 'COMPLETED'   // 已完成
}

/**
 * 报废状态枚举
 */
export enum DisposalStatus {
  PENDING = 'PENDING',      // 待审批
  APPROVED = 'APPROVED',    // 已批准
  REJECTED = 'REJECTED',    // 已拒绝
  COMPLETED = 'COMPLETED'   // 已完成
}

/**
 * 维护类型枚举
 */
export enum MaintenanceType {
  ROUTINE = 'ROUTINE',                  // 例行保养
  REPAIR = 'REPAIR',                    // 维修
  PARTS_REPLACEMENT = 'PARTS_REPLACEMENT', // 部件更换
  CLEANING = 'CLEANING',                // 清洁
  OTHER = 'OTHER'                       // 其他
}

/**
 * 校准结果枚举
 */
export enum CalibrationResult {
  QUALIFIED = 'QUALIFIED',      // 合格
  UNQUALIFIED = 'UNQUALIFIED',  // 不合格
  CONDITIONAL = 'CONDITIONAL'   // 有条件合格
}

// ============================================
// 仪器管理 - 实体接口
// ============================================

/**
 * 技术参数接口
 */
export interface TechnicalParams {
  measurementRange?: string  // 测量范围
  precision?: string         // 精度
  resolution?: string        // 分辨率
  accuracy?: string          // 准确度
  sensitivity?: string       // 灵敏度
  [key: string]: any        // 允许其他自定义参数
}

/**
 * 仪器实体
 */
export interface Instrument {
  id: string
  code: string
  name: string
  model?: string
  manufacturer?: string
  serialNumber?: string
  purchaseDate?: Date
  purchasePrice?: number
  technicalParams?: TechnicalParams
  status: InstrumentStatus
  currentLocation?: string
  currentDepartment?: string
  currentResponsible?: string
  usageYears?: number
  warrantyExpiry?: Date
  description?: string
  remarks?: string
  createdBy: string
  createdAt: Date
  updatedAt: Date
}

/**
 * 仪器流转记录实体
 */
export interface InstrumentTransfer {
  id: string
  instrumentId: string
  fromDepartment: string
  toDepartment: string
  fromResponsible: string
  toResponsible: string
  transferReason?: string
  expectedReturnDate?: Date
  status: InstrumentTransferStatus
  confirmedAt?: Date
  confirmedBy?: string
  rejectedAt?: Date
  rejectedBy?: string
  rejectionReason?: string
  createdBy: string
  createdAt: Date
  updatedAt: Date
}

/**
 * 维护记录实体
 */
export interface MaintenanceRecord {
  id: string
  instrumentId: string
  maintenanceDate: Date
  maintenanceType: MaintenanceType
  maintenanceContent: string
  maintenancePerson: string
  maintenanceCost?: number
  nextMaintenanceDate?: Date
  remarks?: string
  createdBy: string
  createdAt: Date
  updatedAt: Date
}

/**
 * 校准记录实体
 */
export interface CalibrationRecord {
  id: string
  instrumentId: string
  calibrationDate: Date
  calibrationOrg: string
  certificateNumber?: string
  calibrationResult: CalibrationResult
  nextCalibrationDate?: Date
  remarks?: string
  certificateFileId?: string
  createdBy: string
  createdAt: Date
  updatedAt: Date
}

/**
 * 报废记录实体
 */
export interface DisposalRecord {
  id: string
  instrumentId: string
  disposalReason: string
  disposalDate?: Date
  status: DisposalStatus
  approvedBy?: string
  approvedAt?: Date
  rejectedBy?: string
  rejectedAt?: Date
  rejectionReason?: string
  createdBy: string
  createdAt: Date
  updatedAt: Date
}

/**
 * 仪器文档实体
 */
export interface InstrumentDocument {
  id: string
  instrumentId: string
  fileName: string
  fileSize: number
  filePath: string
  fileType: string
  documentType: string  // manual, certificate, photo, report, other
  description?: string
  uploadedBy: string
  uploadedAt: Date
}

/**
 * 维护文档实体
 */
export interface MaintenanceDocument {
  id: string
  maintenanceId: string
  fileName: string
  fileSize: number
  filePath: string
  fileType: string
  description?: string
  uploadedBy: string
  uploadedAt: Date
}

/**
 * 报废文档实体
 */
export interface DisposalDocument {
  id: string
  disposalId: string
  fileName: string
  fileSize: number
  filePath: string
  fileType: string
  description?: string
  uploadedBy: string
  uploadedAt: Date
}

// ============================================
// 仪器管理 - DTO 接口
// ============================================

/**
 * 创建仪器 DTO
 */
export interface CreateInstrumentDto {
  code: string
  name: string
  model?: string
  manufacturer?: string
  serialNumber?: string
  purchaseDate?: Date | string
  purchasePrice?: number
  technicalParams?: TechnicalParams
  status?: InstrumentStatus
  currentLocation?: string
  currentDepartment?: string
  currentResponsible?: string
  usageYears?: number
  warrantyExpiry?: Date | string
  description?: string
  remarks?: string
}

/**
 * 更新仪器 DTO
 */
export interface UpdateInstrumentDto {
  name?: string
  model?: string
  manufacturer?: string
  serialNumber?: string
  purchaseDate?: Date | string
  purchasePrice?: number
  technicalParams?: TechnicalParams
  status?: InstrumentStatus
  currentLocation?: string
  currentDepartment?: string
  currentResponsible?: string
  usageYears?: number
  warrantyExpiry?: Date | string
  description?: string
  remarks?: string
}

/**
 * 仪器查询参数 DTO
 */
export interface InstrumentQueryDto {
  page?: number
  pageSize?: number
  code?: string
  name?: string
  status?: InstrumentStatus
  department?: string
  location?: string
  manufacturer?: string
  search?: string  // 模糊搜索(编码、名称、型号)
  startDate?: Date | string  // 购置日期范围
  endDate?: Date | string
}

/**
 * 仪器响应数据
 */
export interface InstrumentResponse extends Instrument {
  transferCount?: number      // 流转次数
  maintenanceCount?: number   // 维护次数
  calibrationCount?: number   // 校准次数
  documentCount?: number      // 文档数量
  lastMaintenanceDate?: Date  // 最后维护日期
  lastCalibrationDate?: Date  // 最后校准日期
  nextMaintenanceDate?: Date  // 下次维护日期
  nextCalibrationDate?: Date  // 下次校准日期
}

/**
 * 仪器详情响应数据
 */
export interface InstrumentDetailResponse extends InstrumentResponse {
  transfers?: InstrumentTransfer[]
  maintenanceRecords?: MaintenanceRecord[]
  calibrationRecords?: CalibrationRecord[]
  documents?: InstrumentDocument[]
  disposalRecord?: DisposalRecord
}

// ============================================
// 流转管理 - DTO 接口
// ============================================

/**
 * 创建流转申请 DTO
 */
export interface CreateTransferDto {
  instrumentId: string
  fromDepartment: string
  toDepartment: string
  fromResponsible: string
  toResponsible: string
  transferReason?: string
  expectedReturnDate?: Date | string
}

/**
 * 确认流转 DTO
 */
export interface ConfirmTransferDto {
  comments?: string
}

/**
 * 拒绝流转 DTO
 */
export interface RejectTransferDto {
  rejectionReason: string
}

/**
 * 流转查询参数 DTO
 */
export interface TransferQueryDto {
  page?: number
  pageSize?: number
  instrumentId?: string
  status?: InstrumentTransferStatus
  department?: string  // 来源或目标部门
  startDate?: Date | string
  endDate?: Date | string
}

/**
 * 流转响应数据
 */
export interface TransferResponse extends InstrumentTransfer {
  instrument?: {
    id: string
    code: string
    name: string
    model?: string
  }
}

// ============================================
// 维护管理 - DTO 接口
// ============================================

/**
 * 创建维护记录 DTO
 */
export interface CreateMaintenanceDto {
  instrumentId: string
  maintenanceDate: Date | string
  maintenanceType: MaintenanceType
  maintenanceContent: string
  maintenancePerson: string
  maintenanceCost?: number
  nextMaintenanceDate?: Date | string
  remarks?: string
}

/**
 * 更新维护记录 DTO
 */
export interface UpdateMaintenanceDto {
  maintenanceDate?: Date | string
  maintenanceType?: MaintenanceType
  maintenanceContent?: string
  maintenancePerson?: string
  maintenanceCost?: number
  nextMaintenanceDate?: Date | string
  remarks?: string
}

/**
 * 维护查询参数 DTO
 */
export interface MaintenanceQueryDto {
  page?: number
  pageSize?: number
  instrumentId?: string
  maintenanceType?: MaintenanceType
  startDate?: Date | string
  endDate?: Date | string
  maintenancePerson?: string
}

/**
 * 维护响应数据
 */
export interface MaintenanceResponse extends MaintenanceRecord {
  instrument?: {
    id: string
    code: string
    name: string
    model?: string
  }
  documents?: MaintenanceDocument[]
}

/**
 * 维护提醒数据
 */
export interface MaintenanceReminder {
  instrumentId: string
  instrumentCode: string
  instrumentName: string
  lastMaintenanceDate?: Date
  nextMaintenanceDate: Date
  daysUntilMaintenance: number
  isOverdue: boolean
}

// ============================================
// 校准管理 - DTO 接口
// ============================================

/**
 * 创建校准记录 DTO
 */
export interface CreateCalibrationDto {
  instrumentId: string
  calibrationDate: Date | string
  calibrationOrg: string
  certificateNumber?: string
  calibrationResult: CalibrationResult
  nextCalibrationDate?: Date | string
  remarks?: string
  certificateFileId?: string
}

/**
 * 更新校准记录 DTO
 */
export interface UpdateCalibrationDto {
  calibrationDate?: Date | string
  calibrationOrg?: string
  certificateNumber?: string
  calibrationResult?: CalibrationResult
  nextCalibrationDate?: Date | string
  remarks?: string
  certificateFileId?: string
}

/**
 * 校准查询参数 DTO
 */
export interface CalibrationQueryDto {
  page?: number
  pageSize?: number
  instrumentId?: string
  calibrationResult?: CalibrationResult
  startDate?: Date | string
  endDate?: Date | string
  calibrationOrg?: string
}

/**
 * 校准响应数据
 */
export interface CalibrationResponse extends CalibrationRecord {
  instrument?: {
    id: string
    code: string
    name: string
    model?: string
  }
  certificateFile?: InstrumentDocument
}

/**
 * 校准到期提醒数据
 */
export interface CalibrationReminder {
  instrumentId: string
  instrumentCode: string
  instrumentName: string
  lastCalibrationDate?: Date
  nextCalibrationDate: Date
  daysUntilExpiry: number
  isExpired: boolean
}

// ============================================
// 报废管理 - DTO 接口
// ============================================

/**
 * 创建报废申请 DTO
 */
export interface CreateDisposalDto {
  instrumentId: string
  disposalReason: string
  disposalDate?: Date | string
}

/**
 * 批准报废 DTO
 */
export interface ApproveDisposalDto {
  comments?: string
}

/**
 * 拒绝报废 DTO
 */
export interface RejectDisposalDto {
  rejectionReason: string
}

/**
 * 报废查询参数 DTO
 */
export interface DisposalQueryDto {
  page?: number
  pageSize?: number
  status?: DisposalStatus
  startDate?: Date | string
  endDate?: Date | string
}

/**
 * 报废响应数据
 */
export interface DisposalResponse extends DisposalRecord {
  instrument?: {
    id: string
    code: string
    name: string
    model?: string
    purchaseDate?: Date
    purchasePrice?: number
  }
  documents?: DisposalDocument[]
}

// ============================================
// 文档管理 - DTO 接口
// ============================================

/**
 * 上传文档 DTO
 */
export interface UploadDocumentDto {
  fileName: string
  fileSize: number
  filePath: string
  fileType: string
  documentType?: string
  description?: string
}

/**
 * 文档查询参数 DTO
 */
export interface DocumentQueryDto {
  page?: number
  pageSize?: number
  instrumentId?: string
  documentType?: string
  uploadedBy?: string
  startDate?: Date | string
  endDate?: Date | string
}

// ============================================
// 统计分析 - DTO 接口
// ============================================

/**
 * 仪器统计数据
 */
export interface InstrumentStatistics {
  totalCount: number
  statusDistribution: {
    status: InstrumentStatus
    count: number
    percentage: number
  }[]
  departmentDistribution: {
    department: string
    count: number
    percentage: number
  }[]
  manufacturerDistribution: {
    manufacturer: string
    count: number
    percentage: number
  }[]
  averageUsageYears: number
  totalValue: number
}

/**
 * 维护统计数据
 */
export interface MaintenanceStatistics {
  totalCount: number
  typeDistribution: {
    type: MaintenanceType
    count: number
    percentage: number
  }[]
  totalCost: number
  averageCost: number
  upcomingMaintenanceCount: number
  overdueMaintenanceCount: number
}

/**
 * 校准统计数据
 */
export interface CalibrationStatistics {
  totalCount: number
  resultDistribution: {
    result: CalibrationResult
    count: number
    percentage: number
  }[]
  upcomingCalibrationCount: number
  expiredCalibrationCount: number
  qualifiedRate: number
}

/**
 * 流转统计数据
 */
export interface TransferStatistics {
  totalCount: number
  statusDistribution: {
    status: InstrumentTransferStatus
    count: number
    percentage: number
  }[]
  departmentActivity: {
    department: string
    outgoingCount: number
    incomingCount: number
  }[]
}

/**
 * 综合统计数据
 */
export interface ComprehensiveStatistics {
  instrument: InstrumentStatistics
  maintenance: MaintenanceStatistics
  calibration: CalibrationStatistics
  transfer: TransferStatistics
}

// ============================================
// 分页结果类型
// ============================================

/**
 * 分页结果基础接口
 */
export interface PaginatedResult<T> {
  items: T[]
  total: number
  page: number
  pageSize: number
  totalPages: number
}

/**
 * 仪器分页结果
 */
export interface PaginatedInstrumentResult extends PaginatedResult<InstrumentResponse> {}

/**
 * 流转分页结果
 */
export interface PaginatedTransferResult extends PaginatedResult<TransferResponse> {}

/**
 * 维护分页结果
 */
export interface PaginatedMaintenanceResult extends PaginatedResult<MaintenanceResponse> {}

/**
 * 校准分页结果
 */
export interface PaginatedCalibrationResult extends PaginatedResult<CalibrationResponse> {}

/**
 * 报废分页结果
 */
export interface PaginatedDisposalResult extends PaginatedResult<DisposalResponse> {}

/**
 * 文档分页结果
 */
export interface PaginatedDocumentResult extends PaginatedResult<InstrumentDocument> {}
